internal/pkg/cli/env_delete.go

```diff
@@ -5,29 +5,129 @@ package cli
+	iamClient    iamiface.IAMAPI
+	deployClient environmentDeployer
+	prog         progress
+
+	// Cached objects to avoid multiple requests.
+	env *archer.Environment
+
+	*GlobalOpts
 }
 
 // Ask is a no-op for this command.
 func (opts *DeleteEnvOpts) Ask() error {
 	return nil
 }
 
-// Validate returns an error if the environment name does not exist in the project, or there are existing applications
-// in the environment, or the environment is not available using the EnvironmentManagerRole.
+// Validate returns an error if the environment name does not exist in the project, or if there are live applications under the environment.
 func (opts *DeleteEnvOpts) Validate() error {
+	// Validate that the environment is stored in SSM.
+	env, err := opts.storeClient.GetEnvironment(opts.ProjectName(), opts.EnvName)
+	if err != nil {
+		return fmt.Errorf("get environment %s metadata in project %s: %w", opts.EnvName, opts.ProjectName(), err)
+	}
+	opts.env = env
+
+	if opts.rgClient == nil {
+		// Tests mock the client.
+		if err := opts.initClientsFromRole(opts.env.ManagerRoleARN, opts.env.Region); err != nil {
+			return err
+		}
+	}
+
+	// Look up applications by searching for cloudformation stacks in the environment's region.
+	stacks, err := opts.rgClient.GetResources(&resourcegroupstaggingapi.GetResourcesInput{
+		ResourceTypeFilters: []*string{aws.String("cloudformation")},
+		TagFilters: []*resourcegroupstaggingapi.TagFilter{
+			{
+				Key:    aws.String(stack.AppTagKey),
+				Values: []*string{}, // Matches any application stack.
+			},
+			{
+				Key:    aws.String(stack.EnvTagKey),
+				Values: []*string{aws.String(env.Name)},
+			},
+			{
+				Key:    aws.String(stack.ProjectTagKey),
+				Values: []*string{aws.String(env.Project)},
+			},
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("find application cloudformation stacks: %w", err)
+	}
+	if len(stacks.ResourceTagMappingList) > 0 {
+		var appNames []string
+		for _, cfnStack := range stacks.ResourceTagMappingList {
+			for _, t := range cfnStack.Tags {
+				if *t.Key != stack.AppTagKey {
+					continue
+				}
+				appNames = append(appNames, *t.Value)
+			}
+		}
+		return fmt.Errorf("applications: '%s' still exist within the environment %s", strings.Join(appNames, ", "), env.Name)
+	}
 	return nil
 }
 
 // Execute deletes the environment from the project by first deleting the stack and then removing the entry from the store.
+// If an operation fails, it moves on to the next one instead of halting the execution.
+// The environment is removed from the store only if other delete operations succeed.
+// Execute assumes that Validate is invoked first.
 func (opts *DeleteEnvOpts) Execute() error {
+	shouldDelete, err := opts.shouldDelete(opts.env.Project, opts.env.Name)
+	if err != nil {
+		return err
+	}
+
+	if !shouldDelete {
+		return nil
+	}
+	isStackDeleted := opts.deleteStack()
+	areRolesDeleted := opts.deleteRoles()
+	if isStackDeleted && areRolesDeleted {
```

Comment:
We should eventually have a force command or something that will attempt to delete this parameter, even if we can't delete the stacks for some reason.

Reply:
Added TODO